Doc migration

NicolasMassart opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/PegaSysEng/pantheon/blob/master/CONTRIBUTING.md -->

## PR description
Migrate the doc from the wiki to a folder where we will be able to plug Readthedocs.org
## Fixed Issue(s)
<!-- Please link to fixed issue(s) here using format: fixes #<issue number> -->
<!-- Example: "fixes #2" -->
fixes #298 

## How to test this PR ?

**TL;DR**
_You have to build the site locally but if you dont want to there is a preview available at https://pantheon.readthedocs.io/en/latest/ which is based on the branch in my fork for this PR._

The tool used to generate doc is MkDocs.
However, the target will be a Readthedocs.org hosted site. So as RTD requires the doc content to be published in the master branch of the repository, it's not possible to use it to preview the content of the PR.

To overcome this issue, you have to install MkDocs locally (or with a docker image) and use it to compile the doc which will then be served on a local webserver. Refer to [MkDocs installation guide]( https://mkdocs.readthedocs.io/en/stable/#installation ) it's a pretty simple to install Python app, mostly using PIP and the requirement file provided : 

`$ pip install -r /docs/requirements.txt`

This file contains all the dependencies and versions required to build the site. You will have to do this install only once, unless we update the packages.

Then to build and serve the site run: 

`$ mkdocs serve`

You will then see something like : 
```
INFO    -  Building documentation...
INFO    -  Cleaning site directory
INFO    -  The following pages exist in the docs directory, but are not included in the "nav" configuration:
  - index.md
  - GettingStartedBinaries.md
  - Installation/Install-Chocolatey.md
  - Using-Pantheon/FilterExample.md
  - Using-Pantheon/Filters.md
  - Using-Pantheon/Switching from Geth.md
  - architecture/architecture.md
  - global/test_accounts.md
[I 181214 22:53:16 server:298] Serving on http://127.0.0.1:8000
[I 181214 22:53:16 handlers:59] Start watching changes
[I 181214 22:53:16 handlers:61] Start detecting changes
```

The site is then available in your browser at http://127.0.0.1:8000

You can try to change and save a page, it will refresh automatically the site and you will be able to see your change live.

NicolasMassart commented on docs/Configuring-Pantheon/Proof-of-Authority.md:
```diff
@@ -0,0 +1,80 @@
+
+## Genesis File
+
+To use Clique in a private network, Pantheon requires a PoA genesis file. When connecting to Rinkeby, Pantheon uses the [`rinkeby.json`](https://github.com/PegaSysEng/pantheon/blob/master/ethereum/core/src/main/resources/rinkeby.json) genesis file in the `/pantheon/ethereum/core/src/main/resources` directory.
```
Link is not working

NicolasMassart commented on docs/Using-Pantheon/Account-Management.md:
```diff
@@ -0,0 +1,10 @@
+
+Pantheon does not implement private key management. Use third-party tools (for example, [MetaMask](https://consensys.zendesk.com/hc/en-us/articles/360004685212-Generating-MetaMask-Wallet-New-UI-) and [web3j](https://web3j.io/)) for creating accounts. 
+
+In Pantheon, you can use the JSON-RPC methods:
```
empty line needed

NicolasMassart commented on docs/index.md:
```diff
@@ -0,0 +1,26 @@
+title: Pantheon overview
```
Title and desc should be changed as it's what appear when you link to it. So more something like 
title: Pantheon Enterprise Ethereum client
description: Pantheon is an open-source Enterprise Ethereum client developed under the Apache 2.0 license and written in Java. It runs on the Ethereum public network, private networks, and test networks.

MadelineMurray commented:
I'm not sure which updates made it across from the Wiki but at least one I pushed on Friday didn't (Goerli should be in the list of test networks in the Overview). 

The search doesn't do partial string matching if you don't have the start of the string - I can search for `eth_sendRaw` and find `eth_sendRawTransaction` but I can't search for `sendRawTransaction` and find it. 

I think we need to move the Contribute section out of the ReadTheDocs content. I think this belongs on GitHub rather than here. 

_Installation overview_ in left-hand navigation should be _Installation Overview_

Can we reduce the vertical size of the black bar on the bottom of the screen? 

Is it possible to move the version box to directly below the TOC on the left-hand side?  That's where I typically see it on other ReadTheDocs sites.  Although - I'm not sure how often people will be looking for a version other than the latest?  Maybe a questions for Jake? 

I'm not sure about the $ symbols in the code blocks where we've got commands. If you're copying and pasting the commands, you've then got to be careful not to copy the $ as well which is harder than just copying the whole row eg. 
`$ brew tap pegasyseng/pantheon`

I really like the Attention/Notes/Important boxes. We probably need to define when we're using each - on https://pantheon.readthedocs.io/en/latest/Installation/Build-From-Source/ it doesn't feel consistent that _Pantheon requires Java 8+ to compile; earlier versions are not supported._ is Important and _Pantheon is currently supported only on 64-bit versions of Windows, and requires a 64-bit version of JDK/JRE. We recommend that you also remove any 32-bit JDK/JRE installations._ is Note.  I think anything that's going to stop you from progressing should be at least Important rather than Note. 

Is it possible to remove the _Table of contents_ above the navigation on the right-hand side? I'm not sure it's adding any value.

MadelineMurray commented on docs/Configuring-Pantheon/Proof-of-Authority.md:
```diff
@@ -0,0 +1,80 @@
+
+A PoA genesis file defines properties specific to Clique:
+
+!!! example "Genesis JSON file (stripped)"
```
Is stripped commonly used in this way?  It made me look twice - I assume we're using to to indicate the genesis file is incomplete?

NicolasMassart replied:
yes, well perhaps it's not the right word 🤣

MadelineMurray commented on docs/Configuring-Pantheon/Proof-of-Authority.md:
```diff
@@ -0,0 +1,80 @@
+* [clique_discard](../Reference/JSON-RPC-API-Methods.md#clique_discard)
+
+To propose adding a signer, call `clique_propose` specifying the address of the proposed signer and `true`.
+!!! example "JSON-RPC clique_propose request example"
```
These look like titles in the output so I think we should probably be using title case per our doc style guide: 
JSON-RPC clique_propose request example -> JSON-RPC clique_propose Request Example

@donntrenton - what do you think?

NicolasMassart replied:
fixing. I just don't get this title thing about putting first letter uppercase... but I don't get it in french too ;) So I have to remind that I can't spot it directly.

MadelineMurray commented on docs/Using-Pantheon/Transactions.md:
```diff
@@ -0,0 +1,158 @@
+
+These examples describe how to create a signed raw transaction that can be passed to [`eth_sendRawTransaction`](../Reference/JSON-RPC-API-Methods.md#eth_sendrawtransaction).
+
+tip!!!
```
This isn't displaying as a tip box in the output

NicolasMassart replied:
of course I mistyped it, exclamation marks must be at the beginning : `!!!tip` not `tip!!!`

MadelineMurray commented on docs/Using-Pantheon/RPC-PubSub.md:
```diff
@@ -0,0 +1,253 @@
+Subscribing to some events (for example, logs) can cause a flood of notifications while the node is synchronizing.
+
+Notifications are stored in an internal buffer to be sent to the client. 
+If the client falls behind and the number of buffered notifications reaches the limit of ==10,000==, the connection is closed.
```
Same comment as earlier - the yellow highlight really draws my sight to this first and I'm not sure that's actually helpful.

NicolasMassart replied:
Feel free to propose a way to format these datas, given that we now have a lot of means to do it and not just back-ticks that we should keep only for code.

MadelineMurray commented on mkdocs.yml:
```diff
@@ -0,0 +1,121 @@
+    - Contribute:
+        - Code review: community/code-reviews.md
+        - Community membership: community/community-membership.md
+    - References:
```
I think typically is would be 'Reference' rather than 'References'.  The plural version makes me imagine I'm going to get a list of references to other information.

MadelineMurray commented on docs/Reference/JSON-RPC-API-Methods.md:
````diff
@@ -0,0 +1,1911 @@
+    }
+    ```
+
+## Web3 Methods
````
I know this isn't related to your changes but maybe we should take this chance to put the methods into alphabetical order rather than the reasonably random order they're currently in?  Or group the eth_* methods into the methods that are likely to be used together (eg, the ones related to filtering).  If you think grouping is going to be better, we should probably raise a seperate issue for that because that's really more than a formatting change.

NicolasMassart replied:
This will be a nice exercice for my beloved writers ;)

MadelineMurray commented on docs/Reference/JSON-RPC-API-Methods.md:
```diff
@@ -0,0 +1,1911 @@
+
+* `startingBlock` : *quantity* - Index of the highest block on the blockchain when the network synchronization starts.
+
+  If you start with an empty blockchain, the starting block is the beginning of the blockchain (`startingBlock` = 0).
```
Needs indenting in line with the bullet point text

MadelineMurray commented on docs/Reference/JSON-RPC-API-Methods.md:
```diff
@@ -0,0 +1,1911 @@
+
+  If you start with an empty blockchain, the starting block is the beginning of the blockchain (`startingBlock` = 0).
+
+  If you import a block file using `pantheon import <block-file>`, the synchronization starts at the head of the blockchain, and the starting block is the next block synchronized. For example, if you imported 1000 blocks, the import would include blocks 0 to 999, so in that case `startingBlock` = 1000.
```
Needs indenting in line with the bullet point text

MadelineMurray commented on docs/Reference/JSON-RPC-API-Methods.md:
````diff
@@ -0,0 +1,1911 @@
+    ```
+    
+    ```bash tab="wscat WS request"
+    $ wscat -c <JSON-RPC-ws-endpoint:port>
````
I don't think we need these two lines showing the wscat connection - the example request should really be the part they can copy and paste

NicolasMassart replied:
ok

MadelineMurray commented on docs/Reference/JSON-RPC-API-Methods.md:
```diff
@@ -0,0 +1,1911 @@
+
+Returns the current client version.
+
+*Parameters*
```
This is bold for admin_peers - I think it's better bold when scanning the ref.  This comment applies to all of the methods (ie, I think Parameters should be changed to bold throughout)

NicolasMassart replied:
my mistake

MadelineMurray commented on docs/Reference/JSON-RPC-API-Methods.md:
```diff
@@ -0,0 +1,1911 @@
+
+None
+
+*Returns*
```
This is bold for admin_peers - I think it's better bold when scanning the ref. This comment applies to all of the methods (ie, I think Returns should be changed to bold throughout)

NicolasMassart replied:
same

MadelineMurray commented on docs/Reference/JSON-RPC-API-Methods.md:
````diff
@@ -0,0 +1,1911 @@
+    }
+    ```
+
+### eth_getStorageAt
````
There's a paragraph mark after this heading in the output?

MadelineMurray replied:
Actually - I've just realised that's the perma link which is a good feature.  Can we change the symbol it uses to the be the link symbol?  Or is this out of our control?

NicolasMassart replied:
Fixed with fontawesome web font https://fontawesome.com/icons/link?style=solid

MadelineMurray commented on docs/Reference/JSON-RPC-API-Objects.md:
```diff
@@ -0,0 +1,155 @@
+
+Returned by [eth_getFilterChanges](JSON-RPC-API-Methods.md#eth_getfilterchanges) and [transaction receipt objects](#transaction-receipt-object) can contain an array of log objects.  
+
+| Key   |   Type            |  Value  |  
```
Is it possible to make the left hand column of these object tables wider?  The key are currently wrapping but the right column is really wide.  Not sure how much control we have over md tables though.

NicolasMassart replied:
It's due to the abusive use of code highlighting to emphasise words. Bold would be more effective in this context and the word wrap would be correct. Also try to format the tables so that they are readable in the markdown code.

NicolasMassart commented:
> I'm not sure which updates made it across from the Wiki but at least one I pushed on Friday didn't (Goerli should be in the list of test networks in the Overview).

Yes Friday probably, but earliest too. We'll have to check diff during our peer writing session.

> The search doesn't do partial string matching if you don't have the start of the string - I can search for `eth_sendRaw` and find `eth_sendRawTransaction` but I can't search for `sendRawTransaction` and find it.

I filled a bug in MkDocs : https://github.com/mkdocs/mkdocs/issues/1704

> I think we need to move the Contribute section out of the ReadTheDocs content. I think this belongs on GitHub rather than here.

Removed it from nav

> _Installation overview_ in left-hand navigation should be _Installation Overview_

fixed

> Can we reduce the vertical size of the black bar on the bottom of the screen?

fixed

> Is it possible to move the version box to directly below the TOC on the left-hand side? That's where I typically see it on other ReadTheDocs sites. Although - I'm not sure how often people will be looking for a version other than the latest? Maybe a questions for Jake?

Maybe, I don't know, this is part of what's injected by readthedocs so I can control the styles but not where the html is injected... I can make it smaller though...

> I'm not sure about the $ symbols in the code blocks where we've got commands. If you're copying and pasting the commands, you've then got to be careful not to copy the $ as well which is harder than just copying the whole row eg.
> `$ brew tap pegasyseng/pantheon`

I was trying to be consistent with some other sites practices like : 
* https://www.tecmint.com/linux-commands-cheat-sheet/
* https://www.howtoforge.com/linux-commands/
* https://www.thegeekstuff.com/2010/11/50-linux-commands/
but some others are not doing it, so we just have to be consistant with ourselves. I liked the $ on bash as it allows to show the difference with windows commands at first sight. And I guess users that are running linux or mac command lines are aware of this $ sign indicating that you run a regular user and that # is for when you run as root user... shell basics that should be prerequisites IMO.

> I really like the Attention/Notes/Important boxes. We probably need to define when we're using each - on https://pantheon.readthedocs.io/en/latest/Installation/Build-From-Source/ it doesn't feel consistent that _Pantheon requires Java 8+ to compile; earlier versions are not supported._ is Important and _Pantheon is currently supported only on 64-bit versions of Windows, and requires a 64-bit version of JDK/JRE. We recommend that you also remove any 32-bit JDK/JRE installations._ is Note. I think anything that's going to stop you from progressing should be at least Important rather than Note.

I agree with a section about usage of Admonition features to add in the doc style guide which is being written. Also guide about highlighting and source code feature but also meta data, headings usage, ...

> Is it possible to remove the _Table of contents_ above the navigation on the right-hand side? I'm not sure it's adding any value.

Yes, fixed by overriding toc.html

NicolasMassart commented on mkdocs.yml:
```diff
@@ -0,0 +1,121 @@
+        - Mining: Using-Pantheon/Mining.md
+        - RPC Pub/Sub: Using-Pantheon/RPC-PubSub.md
+        - Debugging Pantheon: Using-Pantheon/Debugging.md
+    - Developement:
```
I'm not sure because when they will see how awesome our doc look like, the won't even go back to Github ;)

NicolasMassart replied:
But removed from the nav for now

NicolasMassart commented on docs/Getting-Started/Private-Network-Quickstart.md:
```diff
@@ -0,0 +1,460 @@
+
+The block explorer displays a summary of the private network:
+
+![Block Explorer](ExplorerSummary.png)
```
It shows on mine and on RTD so perhaps a config issue on your side, we'll see.

NicolasMassart commented on docs/Configuring-Pantheon/Networking.md:
```diff
@@ -0,0 +1,41 @@
+
+## Peer Discovery Port
+
+The [`--p2p-listen`](../Reference/Pantheon-CLI-Syntax.md#p2p-listen) option specifies the host and port on which P2P peer discovery listens. The default is ==127.0.0.1:30303==.
```
No, I just wanted to show that we can highlight things, but clearly a pass by Jake could be nice. I have to find his Github user name as he doesn't show up by his name in search...
